### PR TITLE
Add position() to V2 engine

### PR DESCRIPTION
**Pull request by margarit-h**
Signed-off-by: Margarit Hakobyan <margarith@bitquilltech.com>

### Description
Usage: The syntax POSITION(substr IN str) returns the position of the first occurrence of substring substr in string str. Returns 0 if substr is not in str. Returns NULL if any argument is NULL.

Argument type: STRING, STRING, INTEGER

Return type integer:

(STRING IN STRING) -> INTEGER

Example::

    os> SELECT POSITION('world' IN 'helloworld')
    fetched rows / total rows = 1/1
    +-------------------------------------+
    | POSITION('world' IN 'helloworld')   |
    |-------------------------------------|
    | 6                                   |
    +-------------------------------------+
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java [hunk 1]
```diff
@@ -241,6 +242,15 @@ private DefaultFunctionResolver locate() {
                 TextFunction::exprLocate), INTEGER, STRING, STRING, INTEGER));
   }
 
+  /**
+   * POSITION(substr IN str) returns the position of the first occurrence of a substring in a string.
+   * If the substring is not found within the original string, this function returns 0.
+   */
+  private DefaultFunctionResolver position() {
+    return define(BuiltinFunctionName.POSITION.getName(),
+            impl(nullMissingHandling(TextFunction::exprPosition), STRING, STRING, STRING));
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java:251]: Return value type should be `INTEGER`.

File: core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java [hunk 2]
```diff
@@ -241,6 +242,15 @@ private DefaultFunctionResolver locate() {
                 TextFunction::exprLocate), INTEGER, STRING, STRING, INTEGER));
   }
 
+  /**
+   * POSITION(substr IN str) returns the position of the first occurrence of a substring in a string.
+   * If the substring is not found within the original string, this function returns 0.
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java:247]: Maybe: reference `LOCATE` function and signature (see other functions' docs).

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 3]
```diff
@@ -312,6 +313,10 @@ highlightFunction
     : HIGHLIGHT LR_BRACKET relevanceField (COMMA highlightArg)* RR_BRACKET
     ;
 
+positionFunction
+    : POSITION LR_BRACKET inFunctionsArgs RR_BRACKET
```
Comment (Yury-Fridlyand) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:317]: Does this work?
```suggestion
    : POSITION LR_BRACKET functionArg IN functionArg RR_BRACKET
```
So you can remove `inFunctionsArgs`

**Review by Yury-Fridlyand — APPROVED** [review]

File: integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  @Test
+  public void position_function_with_function_as_arg_test() {
+    String query = "SELECT position(upper(str3) IN str1) FROM %s where str1 LIKE 'BINDING SUPPLIES'";
+    JSONObject response = executeJdbcRequest(String.format(query, TestsConstants.TEST_INDEX_CALCS));
+
+    verifySchema(response, schema("position(upper(str3) IN str1)", null, "integer"));
+    assertEquals(1, response.getInt("total"));
+
+    verifyDataRows(response, rows(15));
+  }
```
Comment (GumpacG) [inline integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java:96]: Can we add an IT test for a case where position is in where?

Comment (margarit-h) [reply integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java:96]: Added in c8cd1bb21deda0de7a89ffa8f92f4187a5b9c750 commit

File: core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -213,6 +214,13 @@ public Expression visitHighlightFunction(HighlightFunction node, AnalysisContext
     return new HighlightExpression(expr);
   }
 
+  @Override
+  public Expression visitPositionFunction(PositionFunction node, AnalysisContext context) {
+    Expression left = node.getLeft().accept(this, context);
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java:219]: can we used a better description than `left` and `right`?  I'd prefer to use something like `substringExpr` or `stringPatternExpr` and then `searchStringExpr`?

File: core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java [hunk 6]
```diff
@@ -288,6 +289,10 @@ public UnresolvedExpression highlight(UnresolvedExpression fieldName,
     return new HighlightFunction(fieldName, arguments);
   }
 
+  public UnresolvedExpression position(UnresolvedExpression left, UnresolvedExpression right) {
+    return new PositionFunction(left, right);
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java:293]: Same comment as above:  can we used a better description than left and right?

File: core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.List;
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import org.opensearch.sql.ast.AbstractNodeVisitor;
+
+
+/**
+ * Expression node of Highlight function.
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java:18]: Please update

File: core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ */
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+@Getter
+@ToString
+public class PositionFunction extends UnresolvedExpression {
+  @Getter
+  private UnresolvedExpression left;
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java:26]: Same comment as above:
Can we used a better description than left and right?

File: core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java [hunk 9]
```diff
@@ -170,6 +170,7 @@ public enum BuiltinFunctionName {
   ASCII(FunctionName.of("ascii")),
   LOCATE(FunctionName.of("locate")),
   REPLACE(FunctionName.of("replace")),
+  POSITION(FunctionName.of("position")),
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java:173]: can you re-order this list alphabetically please?  (just the Text functions)

File: core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java [hunk 10]
```diff
@@ -55,6 +55,7 @@ public void register(BuiltinFunctionRepository repository) {
     repository.register(ascii());
     repository.register(locate());
     repository.register(replace());
+    repository.register(position());
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java:58]: Can you re-order this list alphabetically

File: core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java [hunk 11]
```diff
@@ -241,6 +242,18 @@ private DefaultFunctionResolver locate() {
                 TextFunction::exprLocate), INTEGER, STRING, STRING, INTEGER));
   }
 
+  /**
+   * Returns the position of the first occurrence of a substring in a string.
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java:246]: Can you mention that it starts with position 1, rather than 0? Since 0 is a 'failure'.

File: core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java [hunk 12]
```diff
@@ -377,6 +377,20 @@ void locate() {
         DSL.locate(missingRef, DSL.literal("hello"), DSL.literal(1))));
   }
 
+  @Test
+  void position() {
+    FunctionExpression expression = DSL.position(
+        DSL.literal("world"),
+        DSL.literal("helloworld"));
+    assertEquals(INTEGER, expression.type());
+    assertEquals(6, eval(expression).integerValue());
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java:386]: you need a test that returns `0` (missing substring)

File: core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java [hunk 13]
```diff
@@ -377,6 +377,20 @@ void locate() {
         DSL.locate(missingRef, DSL.literal("hello"), DSL.literal(1))));
   }
 
+  @Test
+  void position() {
+    FunctionExpression expression = DSL.position(
+        DSL.literal("world"),
+        DSL.literal("helloworld"));
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java:384]: can you change this to `"helloworldworld"`?  Then it will also cover the duplicate case.

File: docs/user/dql/functions.rst [hunk 14]
```diff
@@ -2331,6 +2331,31 @@ Example::
...
+
+Argument type: STRING, STRING, INTEGER
+
+Return type integer:
+
+(STRING IN STRING) -> INTEGER
+
+Example::
+
+    os> SELECT POSITION('world' IN 'helloworld')
```
Comment (acarbonetto) [inline docs/user/dql/functions.rst:2350]: can you add the 0 case, like:
```
POSITION('invalid' IN 'helloworld')
```

File: integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+  @Test
+  public void position_function_in_where_clause_test() {
+    String query = "SELECT str2 FROM %s WHERE position(str3 IN str2)=1";
+    JSONObject response = executeJdbcRequest(String.format(query, TestsConstants.TEST_INDEX_CALCS));
+
+    verifySchema(response, schema("str2", null, "keyword"));
+    assertEquals(2, response.getInt("total"));
+
+    verifyDataRows(response, rows("eight"), rows("eleven"));
+  }
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java:107]: we need a test for the `null` case for each parameter where it returns `null`.

File: sql/src/test/java/org/opensearch/sql/sql/parser/AstExpressionBuilderTest.java [hunk 16]
```diff
@@ -328,6 +329,14 @@ public void canBuildQualifiedNameHighlightFunction() {
     );
   }
 
+  @Test
+  public void canBuildPositionFunction() {
+    assertEquals(
+            position(AstDSL.qualifiedName("fieldA"), AstDSL.qualifiedName("fieldB")),
+            buildExprAst("position(fieldA IN fieldB)")
+    );
```
Comment (acarbonetto) [inline sql/src/test/java/org/opensearch/sql/sql/parser/AstExpressionBuilderTest.java:337]: we should test this with string literals too.

**Review by acarbonetto — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java [hunk 17]
```diff
@@ -377,6 +377,20 @@ void locate() {
...
+    assertEquals(INTEGER, expression.type());
+    assertEquals(6, eval(expression).integerValue());
+
+    when(nullRef.type()).thenReturn(STRING);
```
Comment (forestmvey) [inline core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java:388]: Can we split these out into `null` and `missing` tests.

File: core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java [hunk 18]
```diff
@@ -241,6 +242,18 @@ private DefaultFunctionResolver locate() {
...
+   * Returns 0 if substring is not in string.
+   * Returns NULL if any argument is NULL.
```
Comment (forestmvey) [inline core/src/main/java/org/opensearch/sql/expression/text/TextFunction.java:248]: What about if missing used as argument?

**Review by forestmvey — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java (hunk 5 above)
Comment (margarit-h) [reply core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java (hunk 6 above)
Comment (margarit-h) [reply core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java (hunk 7 above)
Comment (margarit-h) [reply core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java (hunk 8 above)
Comment (margarit-h) [reply core/src/main/java/org/opensearch/sql/ast/expression/PositionFunction.java]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java (hunk 9 above)
Comment (margarit-h) [reply core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java:164]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java (hunk 12 above)
Comment (margarit-h) [reply core/src/test/java/org/opensearch/sql/expression/text/TextFunctionTest.java:386]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: docs/user/dql/functions.rst (hunk 14 above)
Comment (margarit-h) [reply docs/user/dql/functions.rst:2350]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java (hunk 15 above)
Comment (margarit-h) [reply integ-test/src/test/java/org/opensearch/sql/sql/PositionFunctionIT.java:107]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: sql/src/test/java/org/opensearch/sql/sql/parser/AstExpressionBuilderTest.java (hunk 16 above)
Comment (margarit-h) [reply sql/src/test/java/org/opensearch/sql/sql/parser/AstExpressionBuilderTest.java:337]: Addressed in 9a6b7b5d8962cff8af410638da5151e9ee6a36b9

File: core/src/test/java/org/opensearch/sql/analysis/NamedExpressionAnalyzerTest.java [hunk 19]
```diff
@@ -48,4 +49,14 @@ void visit_highlight() {
     NamedExpression analyze = analyzer.analyze(alias, analysisContext);
     assertEquals("highlight(fieldA)", analyze.getNameOrAlias());
   }
+
+  @Test
+  void visit_position() {
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/analysis/NamedExpressionAnalyzerTest.java:54]: what happens when you pass it the empty string for the string pattern?  Does it return 1 in all cases or is this a failure?

**Review by acarbonetto — APPROVED** [review]

File: core/src/test/java/org/opensearch/sql/analysis/NamedExpressionAnalyzerTest.java (hunk 19 above)
Comment (margarit-h) [reply core/src/test/java/org/opensearch/sql/analysis/NamedExpressionAnalyzerTest.java:54]: It returns 1 - the same as locate() does.
